### PR TITLE
D3ASIM-2845 Prevent Duplicate area naming in the same Area

### PR DESCRIPTION
**Pull request by Hassan754**
### Description
This is part 2 of the restriction to have duplicate area names inside the same parent area
### References
[D3A web side](https://github.com/gridsingularity/d3a-web/pull/1122)

### CODE REVIEW

File: src/d3a/models/area/__init__.py [hunk 1]
```diff
@@ -109,6 +143,29 @@ def __init__(self, name: str = None, children: List["Area"] = None,
...
+    @property
+    def name(self):
+        return self.__name
+
+    @name.setter
+    def name(self, new_name):
+        if not check_area_name_exists_in_parent_area(self.parent, new_name):
+            self.__name = new_name
+        else:
+            raise AreaException("Area name should be unique inside the same Parent Area")
```
Comment (hannesdiedrich) [inline src/d3a/models/area/__init__.py:167]: Minor comment: AFAIK we dont plan to have an area_renaming functionality, but leave it

File: src/d3a/models/area/__init__.py [hunk 2]
```diff
@@ -109,6 +143,29 @@ def __init__(self, name: str = None, children: List["Area"] = None,
...
+    def children(self):
+        return self.__children
+
+    @children.setter
+    def children(self, new_children: AreaChildrenList):
+        if type(new_children) == list:
+            self.__children = AreaChildrenList(self, new_children)
+        elif type(new_children) == AreaChildrenList:
+            new_children.parent_area = self
+            self.__children = new_children
```
Comment (hannesdiedrich) [inline src/d3a/models/area/__init__.py:156]: you probably mean:
```suggestion
            self.children.append(new_children)
```
right?

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: @hannesdiedrich I made the children attribute protected `__children` , so I'm setting it in here .

Comment (hannesdiedrich) [reply src/d3a/models/area/__init__.py:156]: No wait, I a confused. Why do you need the whole block L146-156?

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: This is the case when we have another list of children and we set it to a different parent , I had to set their parent area in here

**Review by hannesdiedrich — COMMENTED** [review]
Left a comment, and here a general comment:
Looks to me like AreaChildrenList is bit of an overkill for this, but maybe you can convince me that this is really needed.

File: src/d3a/models/area/__init__.py (hunk 2 above)
Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: Let's suppose a parent A has Children A , if we had parent B and we set parent B.children = parentA.children , here I'm taking into consideration that we should change the parent of the list

**Comment by Hassan754** [comment]
> Left a comment, and here a general comment:
> Looks to me like AreaChildrenList is bit of an overkill for this, but maybe you can convince me that this is really needed.

I think this is required as I'm overriding the append and insert methods , in a normal list you cannot check for duplication before appending .

File: src/d3a/models/area/__init__.py (hunk 2 above)
Comment (hannesdiedrich) [reply src/d3a/models/area/__init__.py:156]: IMO this is not a case we will ever have. We only change children of parents, never parents of children. 
(you dont Houses from Community to Community). But maybe the others can comment on that.

Comment (faizan2590) [reply src/d3a/models/area/__init__.py:156]: Thinking from the electrical network perspective, it indeed is a possibility to `change parent of children`. Grid Operator could do so to balance two neighbouring radial feeder or to balance load in between transformers or simply to switch tail-end houses to other feeders because they were experience poor voltage regulation.

TLDR: Its indeed a possibility but I wouldn't be in favour of incorporating the targeted code changes so silently and would make a ticket as well as discuss with research team if any immediate need to do so.

**Review by faizan2590 — COMMENTED** [review]
Left a comment, ping me to review once resolved.

File: src/d3a/models/area/__init__.py (hunk 2 above)
Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: @faizan2590 right , I'm open to opening a new ticket for this matter
@spyrostz can we have your opinion on this

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: @faizan2590 I created a ticket and assigned it to the research team https://gridsingularity.atlassian.net/browse/D3ASIM-3229

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: @faizan2590 https://gridsingularity.atlassian.net/browse/D3ASIM-3229?focusedCommentId=21143

Comment (hannesdiedrich) [reply src/d3a/models/area/__init__.py:156]: @Hassan754 AFAIK, protected variables are only prepended with one underscore, not with two, right?

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:156]: > The double underscore “__” (called “dunder“) is used to make an instance attribute or method private (cannot be accessed from outside the class) — when used as leading dunder. When used as enclosing dunder (e.g. “__init__”) it indicates that it is a special method in Python (called “magic method”).

Comment (spyrostz) [reply src/d3a/models/area/__init__.py:156]: Interesting approach in general, but there are 2 major problems with this:
1. there is no need to keep track of the parent in the children list. Each Area instance has a parent member, which already keeps a reference to the parent of this area (it is None for the root of the tree)
2. I highly doubt that we are ever going to move areas from one parent to another, because this might introduce cycles in the tree, and the whole code operates under the assumption that the graph is acyclic (tree is by definition acyclic). If we introduce a graph with cycles, I am certain that multiple parts of d3a will not work (first and foremost, offer/bid forwarding will endlessly forward the same offers and bids in the graph cycle), therefore if we ever need to implement this, we should do this much more carefully. 
For now I personally do not see a strong reason to implement it, even from the electricity network perspective, it will be much easier and far less error prone to destroy and recreate the area objects rather than swap their references. Therefore I would recommend to delete this property altogether.

File: src/d3a/models/area/__init__.py [hunk 3]
```diff
@@ -59,6 +59,39 @@
...
+    :param name: New name of area
+    :return: boolean
+    """
+    for child in parent_area.children:
+        if child.name == name:
+            return True
+    return False
+
+
+class AreaChildrenList(list):
```
Comment (spyrostz) [inline src/d3a/models/area/__init__.py:77]: I am always intimidated when extending the default Python types, because of any potential side effects that these might have. I think a much better approach would be a validator on the constructor of the class, which already has the `children` member populated. IMO this will be much simpler and less error-prone than a new datatype.

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:77]: @spyrostz The problem I am trying to address is the `append` or `insert` or changing a list's parent ( We can eliminate this option for now ) . So for the formers I don't see any other fix

File: src/d3a/models/area/__init__.py (hunk 2 above)
Comment (Hassan754) [reply src/d3a/models/area/__init__.py]: @spyrostz Removed the property

File: src/d3a/models/area/__init__.py (hunk 3 above)
Comment (spyrostz) [reply src/d3a/models/area/__init__.py:77]: Have a look at the Area class constructor. The children constructor argument is a list of children for this area. Therefore, you can easily scan the list names for duplicates directly in the constructor of the area class, and raise the exception there. There is no need to override the insert / append of the list itself.

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:77]: @spyrostz This is only when creating the Area object , we initialize its children , and when we later do `area.children.append(child)` no constructor call is being made

File: src/d3a/models/area/__init__.py [hunk 4]
```diff
@@ -79,12 +112,13 @@ def __init__(self, name: str = None, children: List["Area"] = None,
...
         self.current_tick = 0
-        self.name = name
+        self.__name = name
         self.throughput = throughput
         self.uuid = uuid if uuid is not None else str(uuid4())
         self.slug = slugify(name, to_lower=True)
         self.parent = None
-        self.children = children if children is not None else []
+        self.children = AreaChildrenList(self, children) if children is not None\
+            else AreaChildrenList(self)
```
Comment (spyrostz) [inline src/d3a/models/area/__init__.py:121]:
```suggestion
        child_names = [c.name for c in children]
        if len(set(child_names)) !=len(child_names):
            raise Exception("Duplicate names in children")
        self.children = children
```
Even better, you can abstract it away on a validation function in order to not have a lot of things in the constructor body. IMO it will be way cleaner than overriding the Python list class.

Comment (Hassan754) [reply src/d3a/models/area/__init__.py:121]: This makes sense to me in the creation case , but again if we try to append a child later , how can we validate the uniqueness of name

Comment (spyrostz) [reply src/d3a/models/area/__init__.py:121]: OK now I understand the rationale behind this, thanks for bearing with me. The validation could be ported to the activate() method of the Area class (which already includes references to the parent) however this will be more messy than the solution you have implemented (because activate will also contain validation code which is irrelevant to the strategy activation). Therefore LGTM, leave it as it is.

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by faizan2590 — APPROVED** [review]
